### ISSUE TITLE
Custom plugin cannot find task by type when running with test kit

### ISSUE THREAD

**Issue by mpeyper** [gradle/gradle#24909]
### Expected Behavior

A custom plugin that works when resolved with maven should also work when added by test kit.

### Current Behavior

The custom plugin fails to find tasks by type (i.e. using `project.tasks.withType(ShadowJar::class.java)`) but only when the plugin is added by test kit (i.e. using `GradleRunner.create().withPluginClasspath()`).

The task can be found by name (i.e. using `project.tasks.named("shadowJar")`, however it throws a runtime error if the task is cast to the task type (i.e. using `project.tasks.named("shadowJar", ShadowJar::class.java)`).

### Context (optional)

I am trying to test some new functionality in a custom plugin that applies some common shadow configuration.  The code I'm trying to run looks like this:

```kts
class MyPlugin : Plugin<Project> {
  override fun apply(project: Project) {
    project.pluginManager.withPlugin("com.github.johnrengelman.shadow") {
      project.tasks.withType<ShadowJar>().configureEach {
        transform(MyTransformer()) // I need the `Task` to be cast as a `ShadowJar` for this to work
      }
    }
  }
}
```

This worked when added it to a project locally, but failed when I tried to extract it to a seperate project and write a test for it using test kit.

### Steps to Reproduce

[Repro](https://github.com/mpeyper/testkit-issue-repro)

```shell
git clone git@github.com:mpeyper/testkit-issue-repro.git
cd testkit-issue-repro
./bin/gradle test
```

### Gradle version

7.6.1

### Your Environment (optional)

```txt
❯ gradle --version

------------------------------------------------------------
Gradle 7.6.1
------------------------------------------------------------

Build time:   2023-02-24 13:54:42 UTC
Revision:     3905fe8ac072bbd925c70ddbddddf4463341f4b4

Kotlin:       1.7.10
Groovy:       3.0.13
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          17.0.6 (Eclipse Adoptium 17.0.6+10)
OS:           Mac OS X 13.3.1 aarch64
```

**Comment by runningcode** [comment]
Did you follow the instructions here on automatic injection with the Gava Gradle development plugin? https://docs.gradle.org/current/userguide/test_kit.html#sub:test-kit-automatic-classpath-injection

**Comment by mpeyper** [comment]
Yes, I believe I have followed it correctly. Is there a specific step you think may have been missed?

The plugin is found and is being applied (some of the expected logs appear in the repro) but there just seems to be something going wrong with finding the tasks by type that I believe should still be findable.

**Comment by runningcode** [comment]
The two projects are different. In one project you add the plugin to the buildscript classpath: https://github.com/mpeyper/testkit-issue-repro/blob/main/test-project-maven-local/build.gradle.kts#L11

in the other one you do not: https://github.com/mpeyper/testkit-issue-repro/blob/main/test-project-testkit/build.gradle.kts#L9

**Comment by mpeyper** [comment]
Yes, that is the point. The plugin gets applied regardless of method, and the task exists in both, but only the one from maven local can find the task by type while the one using test kit classpath injection cannot. I would expect either to work.

**Comment by runningcode** [comment]
If you add the plugin to the buildscript classpath in both projects or neither does it behave the same way?

**Comment by mpeyper** [comment]
If I remove the plugin from the buildscript classpath in `test-project-maven-local` (the **neither** case) I get:

* `test-project-testkit` fails like before
* `test-project-maven-local` fails with

```txt
> Configure project :consumer-project
Evaluating project ':consumer-project' using build file '/Users/mpeyper/development/testkit-issue-repro/test-project-maven-local/consumer-project/build.gradle.kts'.
Could not execute [class org.jetbrains.kotlin.gradle.plugin.statistics.KotlinBuildStatHandler.buildFinished]

FAILURE: Build failed with an exception.

* Where:
Build file '/Users/mpeyper/development/testkit-issue-repro/test-project-maven-local/consumer-project/build.gradle.kts' line: 3

* What went wrong:
Plugin [id: 'com.example.plugin.test'] was not found in any of the following sources:

- Gradle Core Plugins (plugin is not in 'org.gradle' namespace)
- Plugin Repositories (plugin dependency must include a version number for this source)
```

If I add the plugin from the buildscript classpath in `test-project-testkit` (the **both** case) I get:

* `test-project-maven-local` passes like before
* `test-project-testkit` fails with:

```txt
> Configure project :
Evaluating root project 'test-project-testkit' using build file '/Users/mpeyper/development/testkit-issue-repro/test-project-testkit/build.gradle.kts'.
Caching disabled for Kotlin DSL plugin accessors for classpath '3b280ac4f1d9b96f159d256490a81553' because:
  Build cache is disabled
Skipping Kotlin DSL plugin accessors for classpath '3b280ac4f1d9b96f159d256490a81553' as it is up-to-date.
Caching disabled for Kotlin DSL script compilation (Project/TopLevel/stage1) because:
  Build cache is disabled
Kotlin DSL script compilation (Project/TopLevel/stage1) is not up-to-date because:
  No history is available.
The configuration :classpath is both resolvable and consumable. This is considered a legacy configuration and it will eventually only be possible to be one of these.
The configuration :classpath is both consumable and declarable. This combination is incorrect, only one of these flags should be set.
The configuration :classpath is both resolvable and consumable. This is considered a legacy configuration and it will eventually only be possible to be one of these.
The configuration :classpath is both consumable and declarable. This combination is incorrect, only one of these flags should be set.
Resource missing. [HTTP HEAD: https://repo.maven.apache.org/maven2/com/example/plugin/test/plugin/0.0.1/plugin-0.0.1.pom]
Resource missing. [HTTP HEAD: https://jcenter.bintray.com/com/example/plugin/test/plugin/0.0.1/plugin-0.0.1.pom]

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring root project 'test-project-testkit'.
> Could not resolve all files for configuration ':classpath'.
   > Could not find com.example.plugin.test:plugin:0.0.1.
     Searched in the following locations:
       - https://repo.maven.apache.org/maven2/com/example/plugin/test/plugin/0.0.1/plugin-0.0.1.pom
       - https://plugins.gradle.org/m2/com/example/plugin/test/plugin/0.0.1/plugin-0.0.1.pom
     Required by:
         project :
```

**Comment by runningcode** [comment]
For the first failure when removing the plugin from the buildscript classpath:
```
* What went wrong:
Plugin [id: 'com.example.plugin.test'] was not found in any of the following sources:

- Gradle Core Plugins (plugin is not in 'org.gradle' namespace)
- Plugin Repositories (plugin dependency must include a version number for this source)
```

Can you make sure that the plugin marker is published to maven local _and_ that you are searching in maven local for the plugin? It does not appear that maven local is a plugin repository listed for your build.

For the second failure, can you share how you did this? I have a suspicion, that the plugin was not declared using the plugin DSL.

**Comment by mpeyper** [comment]
Sorry, I think there might be some confusion about what the two projects represent as there is nothing surprising to me about the project setup or why those two attempts failed the way they failed.

`test-project-maven-local` publishes the [plugin to maven local](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/src/test/kotlin/MyPluginTest.kt#L13) and [references the `mavenLocal` repository](https://github.com/mpeyper/testkit-issue-repro/blob/main/test-project-maven-local/build.gradle.kts#L3) and adds the classpath like you would if you were consuming the plugin as an external dependency (i.e. [add the classpath](https://github.com/mpeyper/testkit-issue-repro/blob/main/test-project-maven-local/build.gradle.kts#L11), [apply the plugin](https://github.com/mpeyper/testkit-issue-repro/blob/main/test-project-maven-local/consumer-project/build.gradle.kts#L5)). The intention here is to simulate how the plugin would be consumed by a real project if it were published for real. This **works as expected** and produces this log:

```txt
> Configure project :consumer-project
Evaluating project ':consumer-project' using build file '/Users/mpeyper/development/testkit-issue-repro/test-project-maven-local/consumer-project/build.gradle.kts'.
The configuration :consumer-project:classpath is both resolvable and consumable. This is considered a legacy configuration and it will eventually only be possible to be one of these.
The configuration :consumer-project:classpath is both consumable and declarable. This combination is incorrect, only one of these flags should be set.
The configuration :consumer-project:classpath is both resolvable and consumable. This is considered a legacy configuration and it will eventually only be possible to be one of these.
The configuration :consumer-project:classpath is both consumable and declarable. This combination is incorrect, only one of these flags should be set.
-- APPLYING MyPlugin --
-- SHADOW PLUGIN APPLIED --
-- EAGERLY found task ':consumer-project:shadowJar' --
-- LAZILY found task ':consumer-project:shadowJar' --
-- BYNAME found task ':consumer-project:shadowJar' (type class com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar_Decorated --
Caching disabled for Kotlin DSL accessors for project ':consumer-project' because:
  Build cache is disabled
Skipping Kotlin DSL accessors for project ':consumer-project' as it is up-to-date.
Caching disabled for Kotlin DSL script compilation (Project/TopLevel/stage2) because:
  Build cache is disabled
Skipping Kotlin DSL script compilation (Project/TopLevel/stage2) as it is up-to-date.
-- CONFIGURE CONSUMER PROJECT --
```

`test-project-testkit` [does not reference `mavenLocal`](https://github.com/mpeyper/testkit-issue-repro/blob/main/test-project-testkit/build.gradle.kts#L2-L5) and expects [`testkit` to provide the plugin in the classpath](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/src/test/kotlin/MyPluginTest.kt#L39) when running the project in the integration test. The intention of this project is to demonstrate how the integration tests are running the plugin and would run before the plugin is published anywhere. This **does not work as expected**, producing this log:

```txt
> Configure project :consumer-project
Evaluating project ':consumer-project' using build file '/Users/mpeyper/development/testkit-issue-repro/test-project-testkit/consumer-project/build.gradle.kts'.
Caching disabled for Kotlin DSL script compilation (Project/TopLevel/stage1) because:
  Build cache is disabled
Skipping Kotlin DSL script compilation (Project/TopLevel/stage1) as it is up-to-date.
The configuration :consumer-project:classpath is both resolvable and consumable. This is considered a legacy configuration and it will eventually only be possible to be one of these.
The configuration :consumer-project:classpath is both consumable and declarable. This combination is incorrect, only one of these flags should be set.
The configuration :consumer-project:classpath is both resolvable and consumable. This is considered a legacy configuration and it will eventually only be possible to be one of these.
The configuration :consumer-project:classpath is both consumable and declarable. This combination is incorrect, only one of these flags should be set.
-- APPLYING MyPlugin --
-- SHADOW PLUGIN APPLIED --
Caching disabled for Kotlin DSL accessors for project ':consumer-project' because:
  Build cache is disabled
Skipping Kotlin DSL accessors for project ':consumer-project' as it is up-to-date.
Caching disabled for Kotlin DSL script compilation (Project/TopLevel/stage2) because:
  Build cache is disabled
Skipping Kotlin DSL script compilation (Project/TopLevel/stage2) as it is up-to-date.
-- BYNAME found task ':consumer-project:shadowJar' (type class com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar_Decorated --
-- CONFIGURE CONSUMER PROJECT --
```

Notice that the [plugin is applied](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/src/main/kotlin/MyPlugin.kt#L8), the [shadow plugin is also applied](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/src/main/kotlin/MyPlugin.kt#L10) and that [the `ShadowJar` task can be found by name (`shadowJar`)](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/src/main/kotlin/MyPlugin.kt#L18), regardless of whether it is from `maven-local` or `testkit`, but the issue is that the `ShadowJar` task can only be found by type (either [eagerly](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/src/main/kotlin/MyPlugin.kt#L12) or [lazily](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/src/main/kotlin/MyPlugin.kt#L15)) for the `maven-local` project.

In both cases, it is using the [same plugin code](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/src/main/kotlin/MyPlugin.kt), which [uses the plugin DSL](https://github.com/mpeyper/testkit-issue-repro/blob/main/plugin/build.gradle.kts#L22-L29).

My expectation is that the plugin should function the same in the integration tests as it would if consumed by a real project so that I am able to test it before it is published.

**Comment by ljacomet** [comment]
For your repository based test, the setup you use is not idiomatic. You are mixing `buildscript` classpath configuration with usage of the `plugins` block. This is not how a project should be configured for real.

Ideally, you should not have to add dependencies for plugins, simply declare them in the `plugins` block. And you would need to add `mavenLocal` to `settings.pluginManagement.repositories`.
If you are looking to not declare a version in the project, see this [documentation](https://docs.gradle.org/current/userguide/plugins.html#sec:plugin_version_management).

I am not sure what's happening with the TestKit version of the build. It feels like a classloading issue because of the mix between plugin classpath injection and dependency declarations.

Could you update your reproducer to remove all `buildscript` blocks, replacing them by the necessary `settings` variants?

**Comment by mpeyper** [comment]
Thanks @ljacomet, that got it.  When the `buildscript` sections are removed, both the tests pass. I did not mind specifying the versions in build scripts so I went that way instead of using the settings files. I've [pushed up the changes](https://github.com/mpeyper/testkit-issue-repro/commit/3b5200079232191cd948d900181a902fde6a034d) to the repro if you are curious what I did.

I'm not sure why our projects tend to use the `buildscript.dependencies` for plugins.  Note that this method is shown in the shadow docs as an alternative, which is why I had assumed it was an ok way to configure things when using a multiproject setup and not hacing to specify the version all over the place.

I'm not sure if there is anything else here for you. The setup I had worked pretty well for the majority of the plugin development (the plugin where this was noticed does a lot more than just apply this config) and has for many months (maybe years), so it felt off when just `plugin.tasks.withType` didn't work, but I'd completely understand if you wanted to close this as user error because the setup did was not done in an idiomatic way.

In either case, my test is working so I will now endeavour to take these learning back to the original plugin.

**Comment by ljacomet** [comment]
Thanks for testing and confirming.

I'll discuss with the team but we might want to keep this issue as a documentation one, so that others do not suffer from a similar issue.

**Comment by jbartok** [comment]
Thank you for your interest in Gradle!

This is a valid documentation issue that we will address.
